opentelemetry-sdk/src/opentelemetry/sdk/trace/__init__.py

```diff
@@ -1007,6 +1032,13 @@ def __init__(
         if shutdown_on_exit:
             self._atexit_handler = atexit.register(self.shutdown)
 
+        if self._resource and self._limits.max_attribute_size:
+            resource_attributes = self._resource.attributes
+            _clean_attributes(
+                resource_attributes, self._limits.max_attribute_size
+            )
+            self._resource = Resource(resource_attributes)
```

Comment:
Can we apply `_clean_attributes` in the `Resource` constructor instead of here?

Reply:
Limits and Resource instances both are provided to TracerProvider as arguments. If we want to do this inside Resource, we need to be able to pass Limits to Resource constructor and expect users to do it. Another thing we can do is add a method on Resource and call it from here so it'd look like:

```python
self._resource = self._resource._with_limits(self._limits)
```

This will allow us to call `_clean_attributes` from within Resource class but not sure how much better that would be as the end result would be pretty much the same i.e, new instance of resource will be created given it is immutable.

Yet another thing we can do is to not apply these limits to resource attributes for now. In practice these limits are useful for spans where an instrumentation might generate very large fields like storing very long stack traces as attributes. Such issues don't plague resources so I'm completely fine with not applying these limits to resource attributes.